### PR TITLE
Flip binary expression assist

### PR DESCRIPTION
**Pull request by marcogroppo**
Adds an assist that can flip these binary comparison operators: `==`, `!=`, `>`, `>=`, `<`, `<=`.

This is a small extension to the 'flip ==' assist.
In theory we could easily flip ANY binary expression, but I'm not sure it would be a good idea (IMHO we should try not to change the meaning of the expression).

Does it make sense?

### CODE REVIEW

**Comment by detrumi** [comment]
For other binary operators that are commonly safe to flip, there's `+`, `*`, `&&` and `||`.

But maybe we want to flip (almost) all operators anyways? Other assists like flip comma don't preserve the meaning either. The only cases that maybe should be blacklisted would be op-assign operators like `+=`.

File: crates/ra_assists/src/flip_binexpr.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+pub(crate) fn flip_binexpr(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let expr = ctx.node_at_offset::<BinExpr>()?;
+    let lhs = expr.lhs()?.syntax();
+    let rhs = expr.rhs()?.syntax();
+    let op_range = expr.op()?.range();
+    // The assist should be available only if the cursor is on the operator
+    let cursor_in_range = ctx.frange.range.is_subrange(&op_range);
+    // The assist should be available only for these binary operators
+    // (it should not change the meaning of the expression)
+    let allowed_ops = [
```
Comment (matklad) [inline crates/ra_assists/src/flip_binexpr.rs]: Let's make it a module-level constant for readability?

File: crates/ra_assists/src/flip_binexpr.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    ];
+    let op_kind = expr.op_kind()?;
+    if !cursor_in_range || !allowed_ops.iter().any(|o| *o == op_kind) {
+        return None;
+    }
+    let new_op = match op_kind {
+        BinOp::GreaterTest => Some("<"),
+        BinOp::GreaterEqualTest => Some("<="),
+        BinOp::LesserTest => Some(">"),
+        BinOp::LesserEqualTest => Some(">="),
```
Comment (matklad) [inline crates/ra_assists/src/flip_binexpr.rs]: There's some duplication with `allowed_ops`, perhaps we need just a single `flip_op -> Option<str>` function?

**Comment by marcogroppo** [comment]
@detrumi just to explain my reasoning: `&&` and `||` are short-circuiting, so flipping the expression changes its meaning in a subtle way. `+` and `*` can be overloaded and for example in the case of String `+` is not commutative. But this is just me looking for problems where there are none 😂

File: crates/ra_assists/src/flip_binexpr.rs (hunk 2 above)
Comment (marcogroppo) [reply crates/ra_assists/src/flip_binexpr.rs]: The first check disables the assist while this one avoids replacing the operator when it's not needed (for example there is no need to change `==` but `<` should be changed to `>`).

Comment (matklad) [reply crates/ra_assists/src/flip_binexpr.rs]: I think it's simpler to just change `==` to `==` :)

Comment (marcogroppo) [reply crates/ra_assists/src/flip_binexpr.rs]: It's slower! :D BTW if we allow (almost) every binary operator... only a few will need to be replaced

Comment (matklad) [reply crates/ra_assists/src/flip_binexpr.rs]: Slowness should not matter here. But yeah, only a few operators need a flip... Anyway, don't have a strong opinion here

**Comment by marcogroppo** [comment]
Now the assist is applied to (almost) every binary operator

**Comment by matklad** [comment]
Awesome, thanks!

bors r+

**Pull request merged**